### PR TITLE
WIP: TZSP support

### PR DESCRIPTION
**Pull request by Thermi**
Full support tzsp encapsulation protocol (used by Mikrotik devices forr sending of data, can be used to capture DNS packets)

Adds full support for TZSP, tested and works. Metadata from TZSP packet is not used, only the payload of it is. We could use the sender address and port as identification of the data source, but that is not implemented right now.

Having a test suite for this would require storing test packets as binary data to ensure the tzsp parser that is used works. That's a different approach to testing than the other collectors used.

EDIT: I found a panic, please wait for me to fix it.
EDIT: fixed

TODO:
- [ ] test case (proposal by dmachard: reuse test case for dnstap like [here](https://github.com/dmachard/go-dns-collector/blob/main/tests/clientquery_dnstaptcp.py))
- [ ] complete darwin (only casting/converting of timeval struct remains)
- [ ] windows (need complete implementation, some other way to get accurate timestamp)

### CODE REVIEW

File: example-config/use-case-12.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    - name: console
+      stdout:
+        mode: json
+      transforms:
+        normalize:
+          qname-lowercase: true
+
+  routes:
+    - from:
+      - tap
```
Comment (dmachard) [inline example-config/use-case-12.yml:28]: `tap` collector doesn't exist

**Comment by dmachard** [comment]
Thanks for this pull request.
To test this collector, perhaps you can add the binary data file in the `testsdata` folder, and write a python file like for the [dnstap](https://github.com/dmachard/go-dns-collector/blob/main/tests/clientquery_dnstaptcp.py)

File: example-config/use-case-12.yml (hunk 1 above)
Comment (Thermi) [reply example-config/use-case-12.yml]: Thank you, fixed.

**Comment by Thermi** [comment]
OK at least the darwin check is ... weird because that error message is bogus.
https://github.com/dmachard/go-dns-collector/actions/runs/3508812413/jobs/5877484502#step:5:50

Go 1.19 on Darwin has that method: https://pkg.go.dev/net@go1.19?GOOS=darwin#UDPConn.ReadFromUDPAddrPort

**Comment by dmachard** [comment]
@Thermi  I pushed a fix in the master to fix your issue with github actions on macOS.
Can you resync your pull request ?

**Comment by Thermi** [comment]
Thank you, I rebased the PR. Let's see what the CI does now.

**Comment by dmachard** [comment]
Can you rebase your PR with the main branch ? 
It will be great to add the support for TZSP protocol

**Comment by Thermi** [comment]
Sure, the only things missing are windows and MacOS support. Or maybe we just don't add that? What we need are the timestamps at receive time in the NIC, or the kernel.

On February 16, 2023 8:22:24 AM UTC, Denis MACHARD ***@***.***> wrote:
>Can you rebase your PR with the main branch ? 
>It will be great to add the support for TZSP protocol
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/dmachard/go-dns-collector/pull/172#issuecomment-1432694338
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from mobile. Please excuse any typos or brevity

**Comment by dmachard** [comment]
No worries , we can always add the support for Windows and macOS later :)

**Comment by dmachard** [comment]
Thank you for the contribution. Sorry for the rebase, a major refactoring has been done.
Do not hesitate to improve it for the next releases

**Pull request merged**